kyaml/yaml/rnode.go

```diff
@@ -440,31 +440,43 @@ func (rn *RNode) getMetaStringField(fName string) string {
 	if md == nil {
 		return ""
 	}
-	f := md.Field(fName)
-	if f.IsNilOrEmpty() {
+	f := yNodeField(md, fName)
+	if f == nil {
 		return ""
 	}
-	return GetValue(f.Value)
+	return f.Value
 }
 
-// getMetaData returns the RNode holding the value of the metadata field.
+// getMetaData returns the yaml.Node for the metadata field.
 // Return nil if field not found (no error).
-func (rn *RNode) getMetaData() *RNode {
+func (rn *RNode) getMetaData() *yaml.Node {
 	if IsMissingOrNull(rn) {
 		return nil
 	}
-	var n *RNode
 	if rn.YNode().Kind == DocumentNode {
 		// get the content if this is the document node
-		n = NewRNode(rn.Content()[0])
-	} else {
-		n = rn
+		return rn.Content()[0]
 	}
-	mf := n.Field(MetadataField)
-	if mf.IsNilOrEmpty() {
+	mf := rn.YNode()
+	if mf == nil {
+		return nil
+	}
+	return yNodeField(mf, MetadataField)
+}
+
+// yNodeField returns the specified field from the node.
+// Returns nil if the node is not a mapping node, or if
+// the field does not exist.
+func yNodeField(n *yaml.Node, field string) *yaml.Node {
```

Comment:
It looks like the implementation of this pretty much the same as `rnode.Field`, with a couple of checks removed. Given that this function gives us more code to maintain that looks like mostly a duplicate of existing code, I have to ask does this really provide that much of a performance impact over using `rnode.Field`?

Reply:
Thanks for the feedback.

I believe it does provide significant improvement, yes. The root issue that this PR attempts to address is the unnecessary memory allocations that come from calls into GetAnnotations - specifically for the purpose of reducing load from PrevIds. PrevIds is called a large number of times, so adjustments there can yield large benefits.

Specifically, GetAnnotations calls getMetaData, which calls Field against "metadata". Field creates (and returns) a new MapNode which is built from two new RNode instances - three new allocations. Then GetAnnotations calls getMapFromMeta for "annotations". getMapFromMeta then calls Field on "annotations" - three more allocations - and runs VisitFields on it. VisitFields then invokes Field against each field, yielding three allocations for every field under "annotations". getMapFromMeta then unrolls the MapNode and its RNodes to extract the strings.

I'm happy to try a change that's just the restriction to the subset of desired keys, but still drives through the VisitFields, and report back the differences.

Reply:
Allowing field subsets into VisitFields and leveraging that from PrevIds's GetAnnotations call...

https://github.com/kubernetes-sigs/kustomize/compare/master...ephesused:kustomize:optimize-annos-2

Using the same tests from my opening comment:

```
CPU
ds-cpu-optanno2.svg: Duration: 13.67s, Total samples = 14.12s (103.29%)
na-cpu-optanno2.svg: Duration: 176.51s, Total samples = 192.39s (108.99%)
ro-cpu-optanno2.svg: Duration: 26.32s, Total samples = 27.94s (106.14%)

Memory
ds-mem-optanno2.svg: Showing nodes accounting for 5224.68MB, 73.22% of 7135.57MB total
na-mem-optanno2.svg: Showing nodes accounting for 91440.79MB, 86.26% of 106001.39MB total
ro-mem-optanno2.svg: Showing nodes accounting for 13402.09MB, 96.24% of 13925.35MB total
```

The benefits are significant compared to the current master, but only about half of what's in this PR.

I follow the desire to avoid duplication, so just let me know which path you'd like me to take - I'm happy to drop this PR in favor of the above commit, or I'm happy to polish this PR.

Just let me know what path you think is the best approach. Thanks again!